### ISSUE TITLE
Feature request: host language codegen at migrate time (ref / TypeScript interface)

### ISSUE THREAD

**Issue by tkalliom** [knex/knex#3691]
An example from the documentation:
```
knex.select('title', 'author', 'year').from('books')
```
This is simple enough. However, what happens if you happen to type `authoer` or `boooks`? You might miss that typo until you actually run the code. However, the sooner you get feedback about errors, the better – it would be really cool that an erroneous query could be highlighted by your IDE. Now, you could make constants for table and column names, but them maintaining when your schema changes is tedious.

For projects that use Knex to produce their database schemas, the migrations already contain an exhaustive listing of names in the host language. It would be cool if the migration tool could optionally for an input along the lines of:
```
up: knex => knex.schema.createTable('books', table => table.string('title') && table.string('author'))
...
up: knex => knex.schema.table('books', table => table.integer('year'))
```
produce something like this:
```
const tBooks = knex.ref('books');
const fBooks = {
  title: knex.ref('title').withSchema('books'),
  author: knex.ref('author').withSchema('books'),
  year: knex.ref('year').withSchema('books')
};
```
so that you could then write your query as `knex.select(fBooks.title, fBooks.author, fBooks.year).from(tBooks)` and have it checked statically. Of course, there are a lot of ways the definitions could be structured; the above with `tBook` and `fBooks` is just an example.

For TypeScript users, the added value could be even greater, as the corresponding interface and a convenience wrapper could be generated as well:
```
interface books {
  title: string;
  author: string;
  year: number;
}
const Books = knex<books>("books") // Books.select(fBooks.title, fBooks.author, fBooks.year)
```

I can think of two strategies for how a codegen like this could work:
1. it could either maintain a schema description in-memory over each migration and then format it into output files: `knex migrate:codegen` (this could be run without a database connection), or
2. it could just get the schema from the database after running the migration and then print the host language representation of it: `knex migrate:latest --codegen`

**Comment by elhigu** [comment]
I don't think this kind of behavior should be included in knex. Knex migrations doesn't have enough information to do this. To properly write support this would require that some functionality must be written that queries schema information from actual database.

It could be possible if knex would have some kind of schema file, which has complete information how database should look like. That would also allow to generate migrations for updating old schema to new one automatically like some ORMs allows to do.

With knex migrations there is no complete information about database schema in single migration or there might not even be complete information all migrations put together if migrations are written to work on top of some initial database dump (thats actually pretty useful too with big long lasting projects).

If it would work only with single migration file or with files that doesn't use any raw queries in migrations it is too opinionated to be added to knex. 

Better way to implement this is to create a separate library that can introspect real database schema from the server and generate javascript / typescript code so that it can be accesses in application. I suppose simple implementation of this could just take schema dump from DB and parse needed parts for outputting the JS / TS.

That kind of generic tool would be very useful for the whole node community to allow other libraries also to use the information about database schema to add more sophisticated features.

Can be reopened if some convincing plan how this could be supported with knex migrations is added.

**Comment by tkalliom** [comment]
The part about starting from an existing schema I did think about, hence the specifier ”projects that use Knex to produce their database schemas”. Migrations with raw SQL, on the other hand, did not occur to me – so yeah, option 1 is not really viable.

Tools which can introspect database schemas and produce TypeScript interfaces do exist; at least [rmp135/sql-ts](https://github.com/rmp135/sql-ts) and [sweetiq/schemats](https://github.com/sweetiq/schemats). The reason I posted this idea here is the code generator would have to output `knex.ref`s to be fully useful, so the tool could not be fully generic.

Perhaps a flag could be added to one of the above tools to optionally produce identifier strings in addition to interfaces. They could be typed with something like `tBooks: TableName<Books> = "books"; fBooks = {author: "author" as ColumnName<Books>}` where `ColumnName<T> extends keyof T`. Then the knex-specific part could be a light wrapper like `qualifiedReference<T>(table: TableName<T>, column: ColumnName<T>) = knex.ref(column).withSchema(table)`
